### PR TITLE
fix(arc): Per-CID locking. KeyMutex

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,7 +2,9 @@ module github.com/ipfs/go-ipfs-blockstore
 
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/gammazero/kmutex v1.0.0
```
Comment (gammazero) [inline go.mod:5]: This needs to be updated to `v1.0.1`.  The `v1.0.0` tag was pulling in the old version that this was forked from, which is not what you want.

File: go.mod [hunk 2]
```diff
@@ -2,7 +2,9 @@ module github.com/ipfs/go-ipfs-blockstore
...
 	github.com/hashicorp/golang-lru v0.5.4
+	github.com/im7mortal/kmutex v1.0.0 // indirect
```
Comment (gammazero) [inline go.mod:7]: After changing to `v1.0.1` and running `go mod tidy`, this should disappear.

File: arc_cache.go [hunk 3]
```diff
@@ -151,6 +170,8 @@ func (b *arccache) PutMany(bs []blocks.Block) error {
 	}
 	for _, block := range good {
 		b.cacheSize(block.Cid(), len(block.RawData()))
+		// unlock after updating store and cache.
```
Comment (gammazero) [inline arc_cache.go:173]: If `PutMany()` returns error (on line 148), then function returns without unlocking the CIDs. Looks like the error from `PutMany()` should be checked _after_ this unlock look, or run the unlock loop in a deferred function.

**Review by gammazero — CHANGES_REQUESTED** [review]

File: go.mod (hunk 2 above)
Comment (frrist) [reply go.mod]: Done

File: arc_cache.go (hunk 3 above)
Comment (frrist) [reply arc_cache.go]: addressed, thanks.

File: go.mod (hunk 1 above)
Comment (gammazero) [reply go.mod]: The maintainer of the original repo accepted my changes, so now you can just use:
```
github.com/im7mortal/kmutex v1.0.1
```

**Review by gammazero — APPROVED** [review]
LGTM. Please rerun the benchmarks with the latest changes.

**Comment by Stebalien** [comment]
Closing based on https://github.com/ipfs/go-ipfs-blockstore/issues/64#issuecomment-829246042.

**Pull request closed without merging**